### PR TITLE
Foreach Test Refactor: Pointwise, Min/Max-imum

### PR DESCRIPTION
**Pull request by crcrpar**
- rewrite pointwise unittests using `ops` decorator
- rewrite minimum&maximum unittests using `ops` decorator
- enable minimum/maximum fastpath for BFloat16
- remove _test_data method

#58833 

cc: @ptrblck @ngimel

### CODE REVIEW

File: test/test_foreach.py [hunk 1]
```diff
@@ -353,105 +339,64 @@ def test_unary_slowpath(self, device, dtype, op):
...
+            ],
+        )
+        self._minmax_test(op, inputs, True, 1)
+
+    @ops(foreach_minmax_op_db)
+    @dtypes(*torch.testing.get_all_fp_dtypes(include_bfloat16=True, include_half=True))
+    def test_minmax_inf_nan(self, device, dtype, op):
+        inputs = (
+            [
+                torch.tensor([inf], device=device, dtype=dtype),
```
Comment (ngimel) [inline test/test_foreach.py:387]: can you please use `float('inf')` etc here, and remove `torch._six` imports? Also, how is this test different from the previous one?

File: test/test_foreach.py [hunk 2]
```diff
@@ -646,38 +591,33 @@ def test_binary_op_tensors_on_different_devices(self, device, dtype, op):
...
-            except RuntimeError as e:
-                with self.assertRaisesRegex(type(e), re.escape(str(e))):
-                    expected = [native_op(t1, t2, t3) for t1, t2, t3 in zip(tensors1, tensors2, tensors3)]
-            else:
+        foreach_op, foreach_op_ = op.method_variant, op.inplace_variant
+        native_op, native_op_ = op.ref, op.ref_inplace
+        try:
+            actual = foreach_op(tensors1, tensors2, tensors3)
+        except RuntimeError as e:
+            with self.assertRaisesRegex(type(e), re.escape(str(e))):
```
Comment (ngimel) [inline test/test_foreach.py:609]: when do errors happen here? It seems like inputs corresponding to the same op are always on the same device, e.g. the below translates into 
```
[native_op(*_cuda_tensors), native_op(*_cpu_tensors)]
```
which should work. Also, no harm in writing it like this, it's a bit cleaner than zip of previously zipped tensors.

Comment (crcrpar) [reply test/test_foreach.py:609]: > when do errors happen here?

cpu kernels don't support bfloat16 and half, but @ops(foreach_pointwiese_op_db) creates the test cases of these two dtypes. Do you think using `@dtypes` decorator and limiting the used dtypes to, e.g., all_fp_dtypes?

> It seems like inputs corresponding to the same op are always on the same device,...

that sounds way better, thank you.

File: test/test_foreach.py (hunk 1 above)
Comment (crcrpar) [reply test/test_foreach.py:387]: > can you please use `float('inf')` etc here, and remove `torch._six` imports?

Does merge of `test_minmax_inf_nan` and `test_minmax_float_inf_nan` sound good to you?



> Also, how is this test different from the previous one?

The only difference is the use of `@ops` decorator.

File: test/test_foreach.py (hunk 2 above)
Comment (ngimel) [reply test/test_foreach.py:609]: Yeah, if the purpose of this test is to test behavior on the different devices, it's better to do it for a single datatype (e.g. float). But you probably want to actually test that inputs on the different devices through an error, and now you are not doing that.

File: test/test_foreach.py [hunk 3]
```diff
@@ -159,13 +173,7 @@ def _test_binary_op_scalar(self, device, dtype, opinfo, N, scalar, is_fastpath,
...
     @ops(foreach_binary_op_db)
     def test_binary_op_scalar_fastpath(self, device, dtype, op):
-        scalars = (
-            random.randint(1, 10),
-            1.0 - random.random(),
-            True,
-            complex(1.0 - random.random(), 1.0 - random.random()),
-        )
-        for N, scalar in itertools.product(N_values, scalars):
+        for N, scalar in itertools.product(N_values, Scalars):
```
Comment (ngimel) [inline test/test_foreach.py:176]: nice refactor

**Review by ngimel — APPROVED** [review]
Nice work, @crcrpar!

**Pull request closed without merging**